Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,403 @@
+* `count` - in/out parameter which must not be `nullptr`. On input it specifies the size of the the `keys` and `values` buffers. On output it contains the number of entries used from `keys` and `values` buffers - the number of properties returned. If the size of the buffers is too small, the function returns a specific error code and fill the `count` with the number of available properties. If `keys` or `values` is `nullptr` the function ignores the input value of `count` and just returns the number of properties.
+* `keys` - buffer which acts as an array of pointers to buffers with keys for the runtime properties.
+* `values` - buffer which acts as an array of pointer to buffers with values for the runtime properties.
+
+`keys` and `values` store pointers to buffers which are owned by the host context. The caller should make a copy of it if it needs to store it for anything longer than immediate consumption. The lifetime is only guaranteed until any of the below happens:
+  * one of the "run" methods is called on the host context
+  * the host context is closed via `hostfxr_close`
+  * the value or existence of any property is changed via `hostfxr_set_runtime_property_value`
+
+Note that `hostfxr_set_runtime_property_value` can remove or add new properties, so the number of properties returned is only valid as long as no properties were added/removed.
+
+
+### Start the runtime
+
+#### Running an application
+``` C
+int hostfxr_run_app(const hostfxr_handle host_context_handle);
+```
+Runs the application specified by the `hostfxr_initialize_for_app`. It is illegal to try to use this function when the host context was initialized through any other way.
+* `host_context_handle` - handle to the initialized host context.
+
+The function will return only once the managed application exits.
+
+`hostfxr_run_app` cannot be used in combination with any other "run" function. It can also only be called once.
+
+
+#### Getting a delegate for runtime functionality
+``` C
+int hostfxr_get_runtime_delegate(const hostfxr_handle host_context_handle, hostfxr_delegate_type type, void ** delegate);
+```
+Starts the runtime and returns a function pointer to specified functionality of the runtime.
+* `host_context_handle` - handle to the initialized host context.
+* `type` - the type of runtime functionality requested
+  * `load_assembly_and_get_function_pointer` - entry point which loads an assembly (with dependencies) and returns function pointer for a specified static method. This is the new way to load managed components by the native host. The intent is that the native app will call `hostfxr_get_runtime_delegate(handle, load_assembly_and_get_function_pointer, &helper)` and then use the `helper` to load assembly and get a function pointer to a method. So something like `helper(assembly_path, type_name, method_name, &function_ptr)`. The exact signature and behavior of this runtime helper is TBD.
```

Comment:
More details needed here.  This is blocking implementation.

Reply:
The details of the runtime helpers should not be part of this document (would make it too big).

I wrote initialize spec for the new runtime helper as part of #5486 , so please look there.

Reply:
Let's move the discussion on the ability to load assemblies and get function pointers to them to #5486.